server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java

```diff
@@ -235,11 +228,11 @@ private void ensureCapacityFromPosition(long newPosition) {
             assert pageSize == newPage.v().length;
             pages.add(newPage);
             // We are at the end of the current page, increment page index
-            if (currentPageOffset == pageSize) {
-                pageIndex++;
-                currentPageOffset = 0;
-            }
             currentCapacity += pageSize;
         }
+        if (currentPageOffset == pageSize) {
```

Comment:
This is also faster for the rare case where we have to allocate multiple pages in a loop to make space here.